src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.cs

```diff
@@ -0,0 +1,535 @@
+        //
+        public override long Length
+        {
+            get
+            {
+                CheckClosed();
+                Debug.Assert(CanSeek);
+
+                long length = 0;
+                for (int pieceNumber = 0; pieceNumber < _dir.GetNumberOfPieces(); ++pieceNumber)
+                {
+                    checked { length += _dir.GetStream(pieceNumber).Length; }
+                }
+                return length;
+            }
+        }
+
+        /// <summary>
+        /// Dispose(bool)
+        /// </summary>
+        /// <param name="disposing"></param>
+        /// <remarks>
+        /// An instance of streams' peculiar dispose pattern, whereby
+        /// the inherited abstract class implements Close by calling
+        /// this virtual protected function.
+        /// In turn, each implementation is responsible for calling back
+        /// its base's implementation.
+        /// </remarks>
+        protected override void Dispose(bool disposing)
+        {
+
+
+            try
+            {
+                if (disposing)
+                {
+                    if (!_closed)
+                    {
+                        _dir.Close();
+                    }
+                }
+            }
+            finally
+            {
+                _closed = true;
+                base.Dispose(disposing);
+            }
+        }
+
+        private void CheckClosed()
+        {
+            if (_closed)
+                throw new ObjectDisposedException(null, SR.StreamObjectDisposed);
+        }
+
+        /// <summary>
+        /// Infer the current piece number from _currentOffset.
+        /// </summary>
+        /// <remarks>
+        /// Storing the current piece number in a field and computing the current offset from it
+        /// would also have been possible, but less efficient.
+        /// </remarks>
+        private int GetCurrentPieceNumber()
+        {
+            // Since this property is likely to be read more often than _currentOffset
+            // gets updated, its value is cached in _currentPieceNumber.
+            // The validity of the cached value is monitored using _offsetForCurrentPieceNumber.
+            if (_offsetForCurrentPieceNumber != _currentOffset)
+            {
+                // Cached value is stale. Refresh.
+                _currentPieceNumber = _dir.GetPieceNumberFromOffset(_currentOffset);
+                _offsetForCurrentPieceNumber = _currentOffset;
+            }
+            return _currentPieceNumber;
+        }
+
+        /// <summary>
+        /// Moves a piece stream at position zero to its new absolute position by reading from it.
+        /// </summary>
+        private static void SeekUnderlyingPieceStream(Stream pieceStream, long byteCount)
+        {
+            // need to add a test for all of this
+            const int BufferSize = 4096;
+            long remainingBytes = byteCount;
+            byte[] readBuffer = new byte[BufferSize];
```

Comment:
Use array from pool, and use ReadExactly to avoid looping.

Reply:
I'll change it to use a pooled array, thanks. I don't want to use ReadExactly though, due to memory usage. Using ReadExactly requires that I allocate an array of size `byteCount`, and this could be very large - if an uncompressed part piece won't fit in memory then I think it's better for the method to take a long time than to crash.

Reply:
Looking more broadly at the method: I want the calling of this to be a corner case. It's trying to seek to a specific position in a stream that doesn't support seeking; that's never going to be efficient.

If a ZipArchiveEntry has a compression method of Stored, the underlying stream will be a SubReadStream. This doesn't support seeking right now, for some reason - I've got a PR in mind to add that. DeflateStreams are by definition, much harder though.

One way to support ad-hoc seeks into a DeflateStream might be to cache X position mappings from the uncompressed to the compressed content as we read, at intervals of `UncompressedLength / X`. A seek could set the position to the best position mapping and read forwards from there, or there could be a `TrySeek` method on a decompressing stream; writing would invalidate relevant parts of the cache. The same concept might also apply to other decompression methods.

This was the point where I decided to keep the PR on track... but this method's got a high performance penalty; I'd like to see it lowered, but it requires upstream changes of some kind.